python/pylibcudf/pylibcudf/column.pyx

```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
             return False
         cumulative_stride *= dim
     return True
+
+
+cdef inline bint _is_valid(const uint8_t* null_bitmap, size_t bit_index):
+    # Example:
+    # - column = [2, None, 3, 1, None, 5, None, 7, 8, None]
+    # - index  = [0, 1,    2, 3, 4,    5, 6,    7, 8, 9]
+    #
+    # Row values:
+    # - row 0 -> valid -> 1
+    # - row 1 -> null  -> 0
+    # - row 2 -> valid -> 1
+    # - row 3 -> valid -> 1
+    # - row 4 -> null  -> 0
+    # - row 5 -> valid -> 1
+    # - row 6 -> null  -> 0
+    # - row 7 -> valid -> 1
+    # - row 8 -> valid -> 1
+    # - row 9 -> null  -> 0
+    #
+    # Build the bitmap:
+    # - index     0,1,2,3,4,5,6,7       76543210 (row_index -> bit_index)
+    # - Byte 0 = [1,0,1,1,0,1,0,1] -> 0b10101101
+    # - Byte 1 = [1,0,0,0,0,0,0,0] -> 0b00000001
+    #
+    # Check the 7th element:
+    # row_index = 6
+    # - row 5 -> valid -> 1
+    # - row 6 -> null  -> 0
+    # - row 7 -> valid -> 1
+    # - row 8 -> valid -> 1
+    # - row 9 -> null  -> 0
+    #
+    # Build the bitmap:
+    # - index     0,1,2,3,4,5,6,7       76543210 (row_index -> bit_index)
+    # - Byte 0 = [1,0,1,1,0,1,0,1] -> 0b10101101
+    # - Byte 1 = [1,0,0,0,0,0,0,0] -> 0b00000001
+    #
+    # Check the 7th element:
+    # row_index = 6
+    # offset = 0
+    # bit_index = offset + row_index = 6
+    # 1. Choose the byte
+    #    bit_index >> 3 = 0
+    #    null_bitmap[0] = 0b10101101
+    # 2. Get the bit position within that byte (bit_index mod 8)
+    #    bit_index & 7 = 6 & 7 = 6
+    # 3. Shift the target bit to the least significant bit position
+    #    0b10101101 >> 6 = 0b00000010
+    # 4. Remove all other bits
+    #    0b00000010 & 1 = 0 (=> the 7th element is null)
+    if null_bitmap == NULL:
+        return 1
+    return ((null_bitmap[bit_index >> 3] >> (bit_index & 7)) & 1) != 0
+
+
+cdef inline void _set_item(list out, int64_t i, object obj) except *:
+    # https://docs.python.org/3/c-api/list.html#c.PyList_SetItem
+    Py_INCREF(obj)
+    PyList_SetItem(out, i, obj)
+
+
+cdef list _arrow_to_pylist_impl(type_id dtype, ArrowArray* arr):
+    cdef size_t n = <size_t>arr.length
+    cdef list out = <list>PyList_New(n)
+    if out is None:
+        raise MemoryError("Unable to create new python list")
+
+    if n == 0:
+        return out
+
+    cdef const void** buffers = arr.buffers
+    cdef int64_t offset = arr.offset
+    cdef size_t bit_index
+    cdef int64_t i, N = n
+    cdef const uint8_t* null_bm = NULL
+    cdef const char* numeric_base = NULL
+    cdef size_t item_size = 0
+    cdef const uint8_t* bool_base = NULL
+    cdef const int32_t* str_offsets = NULL
+    cdef const char* str_base = NULL
+    cdef bint is_bool = dtype == type_id.BOOL8
+    cdef bint is_string = dtype == type_id.STRING
+    cdef bint check_valid = arr.null_count != 0
+
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
```

Comment:
Are the first two clauses in this condition necessary? My thoughts:
- Can you ever have n_buffers == 1? I think you must always have at least two to hold the bitmask and the data, right? Let me know if I'm missing something there.
- If `buffers[0] == NULL` then `null_bm = buffers[0]` is a no-op since you start off with it as NULL anyway.

If I'm right, you either assign `null_bm` unconditionally but continue preempting the call below with `check_valid` (and this condition just becomes `cdef const uint8_t* null_bm = <const uint8_t*>buffers[0] if check_valid else NULL`).

Reply:
I think you have it right. See
- https://arrow.apache.org/docs/format/Columnar.html#fixed-size-primitive-layout
- https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout

The physical memory layout of bool is fixed size.